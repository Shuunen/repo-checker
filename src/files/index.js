const CheckConfigs = require('./configs')
const CheckReadme = require('./readme')
const CheckPackage = require('./package')
const CheckLicense = require('./license')
const CheckTravis = require('./travis')
const CheckRenovate = require('./renovate')
const CheckEditorConfig = require('./editor-config')
const CheckNvmRc = require('./nvmrc')

module.exports = [CheckConfigs, CheckNvmRc, CheckReadme, CheckPackage, CheckLicense, CheckTravis, CheckRenovate, CheckEditorConfig]
