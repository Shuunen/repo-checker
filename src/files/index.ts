export * from './configs'
export * from './editor-config'
export * from './eslint'
export * from './gh-workflow'
export * from './license'
export * from './nvmrc'
export * from './package'
export * from './readme'
export * from './renovate'
export * from './repo-checker'
export * from './travis'
// dam lint
