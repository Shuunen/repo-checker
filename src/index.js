#!/usr/bin/env node

const arg = require('arg')
const path = require('path')

const check = require('./check')
const log = require('./logger')

function init () {
  const args = arg({ '--target': String, '--data': String, '--fix': Boolean }, { argv: process.argv.slice(2) })
  let target = args['--target']
  const doFix = args['--fix']
  if (!target) {
    log.info('\n', `no target specified via : --target=path/to/directory`)
    log.info(`targeting current directory...`, '\n')
    target = '.'
  }
  let data = {}
  if (args['--data']) {
    const p = path.join(__dirname, '..', args['--data'])
    log.info('loading data from', p)
    data = require(p)
  } else if (doFix) {
    log.warn('you should provide data to enhance fix')
  }
  log.start()
    .then(() => check(target, data, doFix))
    .catch(err => log.error(err))
    .then(() => log.end())
}

init()
