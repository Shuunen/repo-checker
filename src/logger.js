/* eslint-disable no-console */
const fs = require('fs')
const chalk = require('chalk')

const pkg = require('../package.json')

class Logger {
  get indent () {
    return '    '.repeat(this.indentLevel)
  }

  get date () {
    return new Date().toISOString().replace('T', ' ').split('.')[0]
  }

  constructor (filePath) {
    this.file = fs.createWriteStream(filePath, { flags: 'a' })
    this.indentLevel = 0
  }

  async _write (...args) {
    this.file.write(args.join(' ') + '\n')
  }

  async setIndentLevel (level) {
    this.indentLevel = level
  }

  debug (...args) {
    return this._write(this.indent, 'üîπ', ...args)
  }

  info (...args) {
    console.log(this.indent, '‚¨ú ', ...args)
    return this._write(this.indent, '‚¨ú', ...args)
  }

  error (...args) {
    console.error(chalk.redBright(this.indent, '‚ùå ', ...args))
    return this._write(this.indent, '‚ùå', ...args)
  }

  warn (...args) {
    console.log(chalk.yellowBright(this.indent, '‚ö†Ô∏è ', ...args))
    return this._write(this.indent, '‚ö†Ô∏è', ...args)
  }

  success (...args) {
    console.log(chalk.green(this.indent, '‚úîÔ∏è ', ...args))
    return this._write(this.indent, '‚úîÔ∏è', ...args)
  }

  test (ok, msg, justWarn) {
    if (ok) {
      this.success(msg)
    } else if (justWarn) {
      this.warn(msg)
    } else {
      this.error(msg)
    }
  }

  fix (...args) {
    console.log(chalk.blueBright(this.indent, '‚¨ú ', ...args))
    return this._write(this.indent, '‚¨ú', ...args)
  }

  line () {
    console.log('')
    return this._write('')
  }

  start () {
    this.line()
    this._write(`‚¨áÔ∏è--- Entry from ${this.date} ---‚¨áÔ∏è`)
    this.info(`${pkg.name} v${pkg.version} is starting`)
    return this.line()
  }

  end () {
    this.line()
    this.info(pkg.name, 'has finished')
    return this.line()
  }
}

const log = new Logger(pkg.config.logFile)

module.exports = log
