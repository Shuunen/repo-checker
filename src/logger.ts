import { bgBlue, black, Logger } from 'shuutils'

class ExtendedLogger extends Logger {
  public fix (...stuff: readonly string[]): void {
    const items = Array.from(stuff)
    items.push(bgBlue(black('[ fixed ]')))
    this.info(...items)
  }

  public unknownError (error: unknown): void {
    if (error instanceof Error) { this.error(error.message); return }
    if (typeof error === 'string') { this.error(error); return }
    this.error('Unknown error format, could not log it')
  }
}

export const log = new ExtendedLogger()
