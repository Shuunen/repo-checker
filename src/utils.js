const fs = require('fs')
const path = require('path')
const log = require('./logger')
const Git = require('simple-git/promise')

async function isGitFolder (folderPath) {
  if (!fs.statSync(folderPath).isDirectory()) {
    return false
  }
  return Git(folderPath).checkIsRepo().then(isRepo => {
    log.debug(`${isRepo ? 'is' : 'NOT'} a git repo : ${folderPath}`)
    return isRepo
  }).catch(err => {
    log.error(`failed at testing git repo : ${folderPath}`, err)
    throw err
  })
}

async function getGitFolders (folderPath) {
  if (await isGitFolder(folderPath)) {
    return [folderPath]
  }
  return new Promise((resolve, reject) => {
    fs.readdir(folderPath, async (err, filePaths) => {
      if (err) {
        log.error(`failed at reading dir : ${folderPath}`, err)
        reject(err)
      }
      let gitDirectories = []
      for (const filePath of filePaths) {
        const p = path.join(folderPath, filePath)
        if (await isGitFolder(p)) {
          gitDirectories.push(p)
        }
      }
      resolve(gitDirectories)
    })
  })
}

function augmentData (folderPath, dataSource) {
  const data = Object.assign({}, dataSource)
  return Git(folderPath).getRemotes(true).then(remotes => {
    const matches = JSON.stringify(remotes).match(/([\w-]+)\/([\w-]+)\.git/)
    if (matches.length === 3) {
      data.user_id = matches[1]
      data.user_id_lowercase = data.user_id.toLowerCase()
      data.repo_id = matches[2]
    }
    return data
  })
}

function folderContainsFile (folderPath, fileName) {
  if (!folderPath || !fileName) {
    return log.error('folderContainsFile miss arguments')
  }
  return new Promise((resolve, reject) => {
    fs.access(path.join(folderPath, fileName), fs.F_OK, (err) => {
      if (err) {
        resolve(false)
      }
      resolve(true)
    })
  })
}

function createFile (folderPath, fileName, fileContent) {
  return new Promise((resolve, reject) => {
    fs.writeFile(path.join(folderPath, fileName), fileContent, 'utf8', (err) => {
      if (err) {
        log.error(err)
        resolve(false)
      }
      resolve(true)
    })
  })
}

function readFile (folderPath, fileName, returnEmptyIfNotExists) {
  return new Promise((resolve, reject) => {
    fs.readFile(path.join(folderPath, fileName), 'utf8', (err, content) => {
      if (err) {
        if (returnEmptyIfNotExists) {
          resolve('')
        }
        reject(err)
      }
      resolve(content)
    })
  })
}

function fillTemplate (template, data) {
  let str = (typeof template === 'object' ? JSON.stringify(template, null, 2) : template) || ''
  if (!str) {
    return str
  }
  const tokens = str.match(/\{{\s?([^}\s]+)\s?}\}/g)
  if (!tokens) {
    return str
  }
  return tokens.reduce(
    (acc, token) => {
      const key = token.replace(/[{\s}]/g, '')
      let value = data && data[key]
      if (!value) {
        log.error(`please provide variable "${key}"`)
      }
      acc = acc.replace(token, value || '')
      return acc
    },
    str
  )
}

module.exports = { log, getGitFolders, fillTemplate, folderContainsFile, createFile, readFile, augmentData }
