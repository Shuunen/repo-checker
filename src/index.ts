#!/usr/bin/env node

import { Nb } from 'shuutils'
import { log } from './logger'
import { start } from './main'

// eslint-disable-next-line unicorn/prefer-top-level-await
start().catch((error: unknown) => {
  log.unknownError(error)
  process.exit(Nb.One)
})
