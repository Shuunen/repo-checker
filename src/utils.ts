import { access, existsSync, lstatSync, readdir, readdirSync, readFile, rmdirSync, stat, unlinkSync, writeFile } from 'fs'
import path from 'path'
import requireFromString from 'require-from-string'
import { promisify } from 'util'
import { dataDefaults, dataFileName, ProjectData } from './constants'
import { log } from './logger'

const readFileAsync = promisify(readFile)
const statAsync = promisify(stat)
const readDirectoryAsync = promisify(readdir)

export async function isGitFolder (folderPath = ''): Promise<boolean> {
  const stat = await statAsync(folderPath)
  if (!stat.isDirectory()) return false
  return await checkFileExists(path.join(folderPath, '.git', 'config'))
}

export async function getGitFolders (folderPath = ''): Promise<string[]> {
  if (await isGitFolder(folderPath)) return [folderPath]
  const filePaths = await readDirectoryAsync(folderPath)
  const gitDirectories = []
  for (const filePath of filePaths) {
    const p = path.join(folderPath, filePath)
    if (await isGitFolder(p)) gitDirectories.push(p)
  }
  return gitDirectories
}

export async function augmentDataWithGit (folderPath: string, dataSource: ProjectData): Promise<ProjectData> {
  const data = new ProjectData(dataSource)
  const gitConfigContent = await readFileInFolder(path.join(folderPath, '.git'), 'config')
  const matches = gitConfigContent.match(/([\w-]+)\/([\w-]+)\.git/) ?? []
  if (matches.length !== 3) return data
  data.user_id = matches[1]
  data.user_id_lowercase = data.user_id.toLowerCase()
  data.repo_id = matches[2]
  log.debug('found user_id in git config :', data.user_id)
  log.debug('found repo_id in git config :', data.repo_id)
  return data
}

export async function augmentDataWithPackageJson (folderPath: string, dataSource: ProjectData): Promise<ProjectData> {
  const data = new ProjectData(dataSource)
  const content = await readFileInFolder(folderPath, 'package.json')
  if (content.length === 0) return data
  data.package_name = content.match(/"name": "([\w+/@-]+)"/)?.[1] ?? dataDefaults.package_name
  data.license = content.match(/"license": "([\w+-.]+)"/)?.[1] ?? dataDefaults.license
  const author = content.match(/"author": "([\s\w/@-]+)\b[\s<]*([\w-.@]+)?>?"/) ?? []
  if (author.length === 3) {
    data.user_name = author[1]
    data.user_mail = author[2] ?? ''
  }
  data.is_module = content.includes('"type": "module"')
  data.user_id = content.match(/github\.com\/([\w-]+)\//)?.[1] ?? dataDefaults.user_id
  data.user_id_lowercase = data.user_id.toLowerCase()
  if (content.includes('"vue"')) data.use_vue = true
  if (content.includes('typescript')) data.use_typescript = true
  if (content.includes('html') || data.use_vue) data.web_published = true
  if (content.includes('npm publish')) data.npm_package = true
  return data
}

export async function augmentData (folderPath: string, dataSource: ProjectData, shouldLoadLocal = false): Promise<ProjectData> {
  let data = new ProjectData(dataSource)
  data = await augmentDataWithGit(folderPath, data)
  data = await augmentDataWithPackageJson(folderPath, data)
  const localDataExists = shouldLoadLocal ? await checkFileExists(path.join(folderPath, dataFileName)) : false
  if (localDataExists) { // local data overwrite the rest
    const localData = requireFromString(await readFileInFolder(folderPath, dataFileName))
    Object.assign(data, localData)
  }
  return data
}

export async function folderContainsFile (folderPath = '', fileName = ''): Promise<boolean> {
  if (folderPath.length === 0) return log.error('folderContainsFile need a folderPath argument')
  if (fileName.length === 0) return log.error('folderContainsFile need a fileName argument')
  return await checkFileExists(path.join(folderPath, fileName))
}

export async function checkFileExists (filePath = ''): Promise<boolean> {
  return await new Promise(resolve => {
    access(filePath, (error) => {
      if (error !== null) return resolve(false)
      resolve(true)
    })
  })
}

export async function createFile (folderPath = '', fileName = '', fileContent = ''): Promise<boolean> {
  if (folderPath.length === 0) return log.error('createFile need a folderPath argument')
  if (fileName.length === 0) return log.error('createFile need a fileName argument')
  return await new Promise(resolve => {
    writeFile(path.join(folderPath, fileName), fileContent, 'utf8', (error) => {
      if (error !== null) {
        log.error(error.message)
        resolve(false)
      }
      resolve(true)
    })
  })
}

export async function readFileInFolder (folderPath = '', fileName = ''): Promise<string> {
  const filePath = path.join(folderPath, fileName)
  const content = await readFileAsync(filePath, 'utf8').catch(error => console.error(error.message))
  return typeof content === 'string' ? content : ''
}

export async function getFileSizeInKo (filePath = ''): Promise<number> {
  const stat = await statAsync(filePath)
  const size = Math.round(stat.size / 1024)
  log.debug('found that file', filePath, 'has a size of :', `${size}`, 'Ko')
  return size
}

export function fillTemplate (template: string | object, data: { [key: string]: string } = {}): string {
  let string = (typeof template === 'object' ? JSON.stringify(template, undefined, 2) : template)
  if (string.length === 0) return string
  const tokens = string.match(/{{\s?([^\s}]+)\s?}}/g)
  if (tokens === null || tokens.length === 0) return string
  for (const token of tokens) {
    const key = token.replace(/[\s{}]/g, '')
    const value = data[key] ?? ''
    if (value.length === 0) {
      log.warn(`cannot fill variable "${key}"`)
      return ''
    }
    string = string.replace(token, value)
  }
  return string
}

// from https://geedew.com/remove-a-directory-that-is-not-empty-in-nodejs/
export function deleteFolderRecursive (path = ''): void {
  if (!existsSync(path)) return
  readdirSync(path).forEach(function (file) {
    const currentPath = path + '/' + file
    if (lstatSync(currentPath).isDirectory()) deleteFolderRecursive(currentPath)
    else unlinkSync(currentPath)
  })
  rmdirSync(path)
}
