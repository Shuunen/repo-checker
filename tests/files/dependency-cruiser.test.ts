import { test } from 'uvu'
import { equal } from 'uvu/assert'
import { DependencyCruiserFile } from '../../src/files'
import { log } from '../../src/logger'
import { promiseFalse } from '../utils'

test('dependency cruiser config missing file', async function () {
  log.canConsoleLog = false
  log.willLogToFile = false
  const instance = new DependencyCruiserFile()
  instance.checkFileExists = promiseFalse
  await instance.start()
  await instance.end()
  equal(instance.passed, [])
  equal(instance.failed, [])
})
