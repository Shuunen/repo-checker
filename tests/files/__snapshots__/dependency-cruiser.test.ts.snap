// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dependency cruiser config missing file 1`] = `[]`;

exports[`dependency cruiser config missing file 2`] = `[]`;

exports[`dependency cruiser config missing file 3`] = `
[
  "use-dependency-cruiser",
]
`;
